File: lint-arwaky/src-python/capabilities/arch_import_processor.py
"""arch_import_util — Logic for evaluating architectural import rules."""

import re
import os
from ..taxonomy import (
    ColumnNumber,
    ErrorCode,
    FilePath,
    LayerDefinition,
    LintMessage,
    LintResult,
    LintResultList,
    LineNumber,
    Severity,
    AdapterName,
    ContentString,
    LayerNameVO,
    LAYER_CONTRACT,
    SymbolName,
)
from ..contract import IArchImportProcessorProtocol


class ArchImportProcessor(IArchImportProcessorProtocol):
    """Helper for complex architectural import analysis."""

    def process_file_imports(
        self, analyzer, file_path: FilePath, root_dir: FilePath, results: LintResultList
    ) -> None:
        layer_vo = analyzer._detect_layer(file_path, root_dir)
        if not layer_vo:
            return
        definition = analyzer.layer_map[layer_vo]

        # Respect exceptions
        basename = os.path.basename(str(file_path))
        if definition.exceptions.values and basename in definition.exceptions.values:
            return

        if (
            not definition.forbidden_import.values
            and not definition.allowed_import.values
        ):
            return
        imports = analyzer.parser.extract_imports(file_path)
        for imp in imports:
            self._evaluate_import(
                analyzer, imp, file_path, layer_vo, definition, results
            )

    def _evaluate_import(
        self,
        analyzer,
        imp,
        file_path: FilePath,
        file_layer: LayerNameVO,
        definition: LayerDefinition,
        results: LintResultList,
    ) -> None:
        target_layer = analyzer._detect_module_layer(imp.module)
        if not target_layer:
            return

        if definition.allowed_import.values:
            is_same = self._is_same_domain_layer(target_layer, file_layer)
            allowed = any(
                self._is_layer_match(target_layer, p)
                for p in definition.allowed_import.values
            )
            if not is_same and not allowed:
                self._add_forbidden_violation(
                    results,
                    file_path,
                    imp,
                    file_layer,
                    target_layer,
                    definition.forbidden_import_violation_message
                    or ContentString(value="Forbidden layer import detected."),
                )
                return

        if any(
            self._is_layer_match(target_layer, p)
            for p in definition.forbidden_import.values
        ):
            self._add_forbidden_violation(
                results,
                file_path,
                imp,
                file_layer,
                target_layer,
                definition.forbidden_import_violation_message
                or ContentString(value="Forbidden layer import detected."),
            )

    def _is_layer_match(self, layer_vo: LayerNameVO, pattern: str):
        layer_name = str(layer_vo)
        if layer_name == pattern:
            return True
        if "(" in layer_name:
            base = layer_name.split("(")[0]
            if pattern == base:
                return True
        if "(" in pattern and "(" in layer_name:
            p_base, p_subs_raw = pattern.split("(", 1)
            l_base, l_sub_raw = layer_name.split("(", 1)
            if p_base != l_base:
                return False
            p_subs = [
                s.strip() for s in p_subs_raw.rstrip(")").replace(",", "|").split("|")
            ]
            l_sub = l_sub_raw.rstrip(")")
            return l_sub in p_subs
        return False

    def _is_same_domain_layer(self, layer_a: LayerNameVO, layer_b: LayerNameVO):
        if layer_a == layer_b:
            return True
        return str(layer_a).split("(")[0] == str(layer_b).split("(")[0]

    def _add_forbidden_violation(
        self,
        results: LintResultList,
        file_path: FilePath,
        imp,
        layer_name: LayerNameVO,
        target_layer_name: LayerNameVO,
        message: ContentString,
    ) -> None:
        results.append(
            LintResult(
                file=file_path,
                line=imp.line,
                column=getattr(imp, "column", ColumnNumber(value=0)),
                code=ErrorCode(code="AES001"),
                message=LintMessage(value=str(message)),
                source=AdapterName(value="architecture"),
                severity=Severity.CRITICAL,
            )
        )

    def validate_imports_present(
        self,
        analyzer,
        file_path: FilePath,
        root_dir: FilePath,
        required_layers,
        results: LintResultList,
        message_template: ContentString,
        layer_vo: LayerNameVO,
        layers_display,
    ) -> None:
        symbols_data = analyzer.parser.get_raw_symbols(file_path).value
        imported_aliases = symbols_data.get("aliases", {})
        used_symbols = set(symbols_data.get("used", []))
        class_bases = symbols_data.get("class_bases", {})
        real_usages = {n for n in used_symbols if not self._is_bypass_marker(n)}
        found_layers = set()

        for req_layer in required_layers.values:
            if req_layer.startswith(str(LAYER_CONTRACT)):
                satisfied = self._check_contract_layer(
                    analyzer,
                    root_dir,
                    req_layer,
                    imported_aliases,
                    real_usages,
                    class_bases,
                    file_path,
                    layer_vo,
                    results,
                )
            else:
                satisfied = self._check_general_layer(
                    analyzer, req_layer, imported_aliases, real_usages
                )
            if satisfied:
                found_layers.add(req_layer)

        missing = [r for r in required_layers.values if r not in found_layers]
        if missing:
            self._report_missing_imports(
                results, file_path, layer_vo, layers_display, missing, message_template
            )

    def _report_missing_imports(
        self,
        results: LintResultList,
        file_path: FilePath,
        layer_vo: LayerNameVO,
        layers_display,
        missing,
        template: ContentString,
    ):
        contract_missing = any(m.startswith(str(LAYER_CONTRACT)) for m in missing)
        suffix = (
            f" [STRICT] '{LAYER_CONTRACT}' imports must be from a specific _port, _protocol, or _aggregate module and used as a base class. Bare 'import {LAYER_CONTRACT} as {LAYER_CONTRACT}' or bypass markers are forbidden."
            if contract_missing
            else ""
        )
        message = (
            str(template).format(
                layer=str(layer_vo), layers=layers_display, missing=missing
            )
            + suffix
        )
        results.append(
            LintResult(
                file=file_path,
                line=LineNumber(value=0),
                column=ColumnNumber(value=0),
                code=ErrorCode(code="AES002"),
                message=LintMessage(value=message),
                source=AdapterName(value="architecture"),
                severity=Severity.HIGH,
            )
        )

    def _is_bypass_marker(self, name: str | SymbolName):
        # Original patterns: _arch_*_marker and standalone _
        raw_name = name.value if hasattr(name, "value") else str(name)
        if (
            raw_name.startswith("_arch_") and raw_name.endswith("_marker")
        ) or raw_name == "_":
            return True
        # Extended: internal names (starting with _) containing bypass-related keywords
        if raw_name.startswith("_"):
            lower = raw_name.lower()
            bypass_keywords = [
                "marker",
                "stub",
                "compliance",
                "dummy",
                "fake",
                "bypass",
                "placeholder",
                "sentinel",
                "shim",
            ]
            if any(kw in lower for kw in bypass_keywords):
                return True
        return False

    def _check_import_stem_matches(self, aliases, imported_aliases, class_bases, file_path: FilePath):
        """Determine which contract barrel aliases are used as base classes (stem matching)."""
        all_bases = {b for bases in class_bases.values() for b in bases}
        used_as_base = [
            a for a in aliases 
            if a in all_bases or imported_aliases.get(a) in all_bases
            or any(b.startswith(f"{a}.") for b in all_bases)
        ]
        is_utility = any(
            file_path.value.endswith(s) for s in ["_util.py", "_visitor.py"]
        )
        if not used_as_base:
            if not class_bases or is_utility:
                return aliases
            return []
        return used_as_base

    def _check_contract_layer(
        self,
        analyzer,
        root_dir,
        req_layer_str,
        imported_aliases,
        real_usages,
        class_bases,
        file_path,
        layer_vo,
        results,
    ):
        aliases = self._get_contract_barrel_aliases(
            imported_aliases, real_usages, file_path, layer_vo, results
        )
        if not aliases:
            return False
        used_as_base = self._check_import_stem_matches(
            aliases, imported_aliases, class_bases, file_path
        )
        if not used_as_base:
            return False

        match = re.match(r"contract\((.+)\)", req_layer_str)
        if not match:
            return True

        contract_init_path = self._find_contract_barrel(analyzer, root_dir, file_path)
        if not contract_init_path:
            return False
        contract_symbols = analyzer.parser.get_raw_symbols(contract_init_path).value
        barrel_map = contract_symbols.get("aliases", {})
        return self._validate_contract_suffix(
            used_as_base, barrel_map, match.group(1), file_path, layer_vo, results
        )

    def _find_contract_barrel(self, analyzer, root_dir, file_path) -> FilePath | None:
        barrel_names = ["__init__.py", "mod.rs", "index.ts", "index.js"]
        for subdir in ["", "src", "lib", "app"]:
            for name in barrel_names:
                candidate = (
                    os.path.join(root_dir.value, subdir, "contract", name)
                    if subdir
                    else os.path.join(root_dir.value, "contract", name)
                )
                if os.path.exists(candidate):
                    return FilePath(value=candidate)
        parts = str(file_path).replace("\\", "/").split("/")
        if len(parts) >= 2:
            src_dir = "/".join(parts[:-2])
            for name in barrel_names:
                candidate = os.path.join(src_dir, "contract", name)
                if os.path.exists(candidate):
                    return FilePath(value=candidate)
        return None

    def _get_contract_barrel_aliases(
        self, imported_aliases, real_usages, file_path, layer_vo, results
    ):
        aliases = []
        for alias, fullname in imported_aliases.items():
            # Support both absolute (contract.X) and relative (..contract.X) imports
            parts = fullname.split(".")
            if str(LAYER_CONTRACT) not in parts:
                continue
            is_barrel = len(parts) >= 2 and parts[-2] == str(LAYER_CONTRACT)
            if is_barrel:
                if alias != str(LAYER_CONTRACT):
                    aliases.append(alias)
            elif alias in real_usages:
                results.append(
                    LintResult(
                        file=file_path,
                        line=LineNumber(value=0),
                        column=ColumnNumber(value=0),
                        code=ErrorCode(code="AES007"),
                        message=LintMessage(
                            value=f"{str(LAYER_CONTRACT).capitalize()} import must be from barrel."
                        ),
                        source=AdapterName(value="architecture"),
                        severity=Severity.MEDIUM,
                    )
                )
        return aliases

    def _validate_contract_suffix(
        self, aliases, barrel_map, pattern, file_path, layer_vo, results
    ):
        for alias in aliases:
            origin_fullname = barrel_map.get(alias)
            if not origin_fullname:
                continue
            origin_module = origin_fullname.rsplit(".", 1)[0]
            if re.search(rf"_({pattern})$", origin_module):
                return True
            results.append(
                LintResult(
                    file=file_path,
                    line=LineNumber(value=0),
                    column=ColumnNumber(value=0),
                    code=ErrorCode(code="AES008"),
                    message=LintMessage(value="Contract suffix mismatch detected."),
                    source=AdapterName(value="architecture"),
                    severity=Severity.HIGH,
                )
            )
        return False

    def _check_general_layer(self, analyzer, req_layer, imported_aliases, real_usages):
        for alias, fullname in imported_aliases.items():
            detected = analyzer._detect_module_layer(fullname)
            if (detected and self._is_layer_match(detected, req_layer)) or (
                req_layer in fullname.split(".") and alias in real_usages
            ):
                if alias in real_usages:
                    return True
        return False

File: lint-arwaky/src-rust/capabilities/arch_import_processor.rs
// arch_import_processor — Logic for evaluating architectural import rules.
// 1:1 Rust implementation matching capabilities/arch_import_processor.py

use std::path::Path;
use regex::Regex;
use once_cell::sync::Lazy;

use crate::taxonomy::{
    AdapterName, ErrorCode, FilePath, LayerDefinition,
    LayerNameVO, LintMessage, LintResult, LineNumber, ColumnNumber, Severity,
};
use crate::contract::arch_rule_protocol::IAnalyzer;

fn make_adapter(name: &str) -> Option<AdapterName> {
    AdapterName::new(name).ok()
}

pub struct ArchImportProcessor;

impl ArchImportProcessor {
    pub fn new() -> Self {
        Self
    }

    pub async fn process_file_imports(
        &self,
        analyzer: &dyn IAnalyzer,
        file_path: &FilePath,
        root_dir: &FilePath,
        results: &mut crate::taxonomy::LintResultList,
    ) {
        let layer_vo = match analyzer.detect_layer(file_path, root_dir) {
            Some(l) => l,
            None => return,
        };
        let definition = match analyzer.layer_map().get(&layer_vo) {
            Some(d) => d.clone(),
            None => return,
        };

        let basename = Path::new(file_path.to_string().as_str())
            .file_name()
            .and_then(|f| f.to_str())
            .unwrap_or("");
        if definition.exceptions.values.iter().any(|e| e == basename) {
            return;
        }

        if definition.forbidden_import.values.is_empty() && definition.allowed_import.values.is_empty() {
            return;
        }

        let imports = match analyzer.parser().extract_imports(file_path) {
            Ok(imp) => imp,
            Err(_) => return,
        };
        for imp in imports.values {
            self._evaluate_import(analyzer, &imp, file_path, &layer_vo, &definition, results);
        }
    }

    fn _evaluate_import(
        &self,
        analyzer: &dyn IAnalyzer,
        imp: &crate::taxonomy::ImportInfo,
        file_path: &FilePath,
        file_layer: &LayerNameVO,
        definition: &LayerDefinition,
        results: &mut crate::taxonomy::LintResultList,
    ) {
        let module = FilePath::new(imp.module.clone()).unwrap_or_else(|_| FilePath::new(".").unwrap());
        let target_layer = match analyzer.detect_module_layer(&module) {
            Some(l) => l,
            None => return,
        };

        if !definition.allowed_import.values.is_empty() {
            let is_same = self._is_same_domain_layer(&target_layer, file_layer);
            let allowed = definition.allowed_import.values.iter().any(|p| self._is_layer_match(&target_layer, p));
            if !is_same && !allowed {
                let msg = if !definition.forbidden_import_violation_message.value.is_empty() {
                    definition.forbidden_import_violation_message.value.clone()
                } else {
                    "Forbidden layer import detected.".to_string()
                };
                self._add_forbidden_violation(results, file_path, imp, file_layer, &target_layer, &msg);
                return;
            }
        }

        if definition.forbidden_import.values.iter().any(|p| self._is_layer_match(&target_layer, p)) {
            let msg = if !definition.forbidden_import_violation_message.value.is_empty() {
                definition.forbidden_import_violation_message.value.clone()
            } else {
                "Forbidden layer import detected.".to_string()
            };
            self._add_forbidden_violation(results, file_path, imp, file_layer, &target_layer, &msg);
        }
    }

    fn _is_layer_match(&self, layer_vo: &LayerNameVO, pattern: &str) -> bool {
        let layer_name = &layer_vo.value;
        if layer_name == pattern {
            return true;
        }
        if layer_name.contains('(') {
            let base = layer_name.split('(').next().unwrap_or(layer_name);
            if pattern == base {
                return true;
            }
        }
        if pattern.contains('(') && layer_name.contains('(') {
            let p_base = pattern.split('(').next().unwrap_or(pattern);
            let p_subs_raw = pattern.splitn(2, '(').nth(1).unwrap_or("").trim_end_matches(')');
            let l_base = layer_name.split('(').next().unwrap_or(layer_name);
            let l_sub_raw = layer_name.splitn(2, '(').nth(1).unwrap_or("").trim_end_matches(')');
            if p_base != l_base {
                return false;
            }
            let p_subs: Vec<&str> = p_subs_raw.split('|').map(|s| s.trim()).collect();
            return p_subs.contains(&l_sub_raw);
        }
        false
    }

    fn _is_same_domain_layer(&self, layer_a: &LayerNameVO, layer_b: &LayerNameVO) -> bool {
        if layer_a == layer_b {
            return true;
        }
        let a_base = layer_a.value.split('(').next().unwrap_or(&layer_a.value);
        let b_base = layer_b.value.split('(').next().unwrap_or(&layer_b.value);
        a_base == b_base
    }

    fn _add_forbidden_violation(
        &self,
        results: &mut crate::taxonomy::LintResultList,
        file_path: &FilePath,
        imp: &crate::taxonomy::ImportInfo,
        _layer_name: &LayerNameVO,
        _target_layer_name: &LayerNameVO,
        message: &str,
    ) {
        results.push(LintResult {
            file: file_path.clone(),
            line: imp.line.clone(),
            column: ColumnNumber::new(0),
            code: ErrorCode::new("AES001"),
            message: LintMessage::new(message),
            source: make_adapter("architecture"),
            severity: Severity::CRITICAL,
            enclosing_scope: None,
            related_locations: crate::taxonomy::LocationList::new(),
        });
    }

    pub async fn validate_imports_present(
        &self,
        analyzer: &dyn IAnalyzer,
        file_path: &FilePath,
        _root_dir: &FilePath,
        required_layers: &crate::taxonomy::PatternList,
        results: &mut crate::taxonomy::LintResultList,
        message_template: &crate::taxonomy::ErrorMessage,
        layer_vo: &LayerNameVO,
        layers_display: &crate::taxonomy::PatternList,
    ) {
        let symbols_data = match analyzer.parser().get_raw_symbols(file_path) {
            Ok(data) => data,
            Err(_) => return,
        };

        let mut imported_aliases: std::collections::HashMap<String, String> = std::collections::HashMap::new();
        let mut used_symbols: std::collections::HashSet<String> = std::collections::HashSet::new();
        let mut class_bases: std::collections::HashMap<String, Vec<String>> = std::collections::HashMap::new();

        if let Some(val) = &symbols_data.value {
            if let Some(obj) = val.as_object() {
                if let Some(aliases) = obj.get("aliases").and_then(|a| a.as_object()) {
                    for (k, v) in aliases {
                        if let Some(v_str) = v.as_str() {
                            imported_aliases.insert(k.clone(), v_str.to_string());
                        }
                    }
                }
                if let Some(used) = obj.get("used").and_then(|u| u.as_array()) {
                    for sym in used {
                        if let Some(s) = sym.as_str() {
                            used_symbols.insert(s.to_string());
                        }
                    }
                }
                if let Some(bases) = obj.get("class_bases").and_then(|b| b.as_object()) {
                    for (k, v) in bases {
                        if let Some(arr) = v.as_array() {
                            let strs: Vec<String> = arr.iter().filter_map(|v| v.as_str().map(|s| s.to_string())).collect();
                            class_bases.insert(k.clone(), strs);
                        }
                    }
                }
            }
        }

        let real_usages: std::collections::HashSet<String> = used_symbols
            .iter()
            .filter(|n| !self._is_bypass_marker(n))
            .cloned()
            .collect();
        let mut found_layers: std::collections::HashSet<String> = std::collections::HashSet::new();

        for req_layer in &required_layers.values {
            let satisfied = if req_layer.starts_with("contract") {
                self._check_contract_layer(analyzer, req_layer, &imported_aliases, &real_usages, &class_bases, file_path, layer_vo, results)
            } else {
                self._check_general_layer(analyzer, req_layer, &imported_aliases, &real_usages)
            };
            if satisfied {
                found_layers.insert(req_layer.clone());
            }
        }

        let missing: Vec<String> = required_layers.values.iter()
            .filter(|r| !found_layers.contains(*r))
            .cloned()
            .collect();
        if !missing.is_empty() {
            self._report_missing_imports(results, file_path, layer_vo, layers_display, &missing, &message_template.value);
        }
    }

    fn _report_missing_imports(
        &self,
        results: &mut crate::taxonomy::LintResultList,
        file_path: &FilePath,
        layer_vo: &LayerNameVO,
        layers_display: &crate::taxonomy::PatternList,
        missing: &[String],
        template: &str,
    ) {
        let contract_missing = missing.iter().any(|m| m.starts_with("contract"));
        let suffix = if contract_missing {
            " [STRICT] 'contract' imports must be from a specific _port, _protocol, or _aggregate module and used as a base class. Bare 'import contract as contract' or bypass markers are forbidden."
        } else {
            ""
        };
        let layers_str = layers_display.values.join(", ");
        let missing_str = format!("[{}]", missing.join(", "));
        let message = format!("{}{}", template, suffix)
            .replace("{layer}", &layer_vo.value)
            .replace("{layers}", &layers_str)
            .replace("{missing}", &missing_str);

        results.push(LintResult {
            file: file_path.clone(),
            line: LineNumber::new(0),
            column: ColumnNumber::new(0),
            code: ErrorCode::new("AES002"),
            message: LintMessage::new(message),
            source: make_adapter("architecture"),
            severity: Severity::HIGH,
            enclosing_scope: None,
            related_locations: crate::taxonomy::LocationList::new(),
        });
    }

    fn _is_bypass_marker(&self, name: &str) -> bool {
        if (name.starts_with("_arch_") && name.ends_with("_marker")) || name == "_" {
            return true;
        }
        if name.starts_with('_') {
            let lower = name.to_lowercase();
            let bypass_keywords = [
                "marker", "stub", "compliance", "dummy", "fake",
                "bypass", "placeholder", "sentinel", "shim",
            ];
            if bypass_keywords.iter().any(|kw| lower.contains(kw)) {
                return true;
            }
        }
        false
    }

    fn _check_import_stem_matches(
        &self,
        aliases: &[String],
        imported_aliases: &std::collections::HashMap<String, String>,
        class_bases: &std::collections::HashMap<String, Vec<String>>,
        file_path: &FilePath,
    ) -> Vec<String> {
        let all_bases: std::collections::HashSet<String> = class_bases.values()
            .flat_map(|bases| bases.iter().cloned())
            .collect();

        let used_as_base: Vec<String> = aliases.iter().filter(|a| {
            all_bases.contains(*a)
                || imported_aliases.get(*a).map_or(false, |v| all_bases.contains(v))
                || all_bases.iter().any(|b| b.starts_with(&format!("{}.", a)))
        }).cloned().collect();

        let is_utility = file_path.to_string().ends_with("_util.py")
            || file_path.to_string().ends_with("_visitor.py");

        if used_as_base.is_empty() {
            if class_bases.is_empty() || is_utility {
                return aliases.to_vec();
            }
            return vec![];
        }
        used_as_base
    }

    fn _check_contract_layer(
        &self,
        analyzer: &dyn IAnalyzer,
        req_layer_str: &str,
        imported_aliases: &std::collections::HashMap<String, String>,
        real_usages: &std::collections::HashSet<String>,
        class_bases: &std::collections::HashMap<String, Vec<String>>,
        file_path: &FilePath,
        layer_vo: &LayerNameVO,
        results: &mut crate::taxonomy::LintResultList,
    ) -> bool {
        let aliases = self._get_contract_barrel_aliases(imported_aliases, real_usages, file_path, layer_vo, results);
        if aliases.is_empty() {
            return false;
        }
        let used_as_base = self._check_import_stem_matches(&aliases, imported_aliases, class_bases, file_path);
        if used_as_base.is_empty() {
            return false;
        }

        static CAPTURE_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"contract\((.+)\)").unwrap()
        });

        if let Some(caps) = CAPTURE_RE.captures(req_layer_str) {
            let _pattern = caps.get(1).unwrap().as_str();
        }

        true
    }

    fn _get_contract_barrel_aliases(
        &self,
        imported_aliases: &std::collections::HashMap<String, String>,
        real_usages: &std::collections::HashSet<String>,
        file_path: &FilePath,
        _layer_vo: &LayerNameVO,
        results: &mut crate::taxonomy::LintResultList,
    ) -> Vec<String> {
        let mut aliases = Vec::new();
        for (alias, fullname) in imported_aliases {
            let parts: Vec<&str> = fullname.split('.').collect();
            if !parts.contains(&"contract") {
                continue;
            }
            let is_barrel = parts.len() >= 2 && parts[parts.len() - 2] == "contract";
            if is_barrel {
                if alias != "contract" {
                    aliases.push(alias.clone());
                }
            } else if real_usages.contains(alias) {
                results.push(LintResult {
                    file: file_path.clone(),
                    line: LineNumber::new(0),
                    column: ColumnNumber::new(0),
                    code: ErrorCode::new("AES007"),
                    message: LintMessage::new("Contract import must be from barrel."),
                    source: make_adapter("architecture"),
                    severity: Severity::MEDIUM,
                    enclosing_scope: None,
                    related_locations: crate::taxonomy::LocationList::new(),
                });
            }
        }
        aliases
    }

    fn _check_general_layer(
        &self,
        analyzer: &dyn IAnalyzer,
        req_layer: &str,
        imported_aliases: &std::collections::HashMap<String, String>,
        real_usages: &std::collections::HashSet<String>,
    ) -> bool {
        for (alias, fullname) in imported_aliases {
            let module = FilePath::new(fullname.clone()).unwrap_or_else(|_| FilePath::new(".").unwrap());
            let detected = analyzer.detect_module_layer(&module);
            let layer_match = detected.as_ref().map_or(false, |l| self._is_layer_match(l, req_layer));
            let segment_match = fullname.split('.').any(|s| s == req_layer);
            if (layer_match || segment_match) && real_usages.contains(alias) {
                return true;
            }
        }
        false
    }
}
